core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -575,6 +578,7 @@ public PlanOptimizers(
                                 .addAll(columnPruningRules)
                                 .add(new InlineProjections(plannerContext, typeAnalyzer))
                                 .addAll(new PushFilterThroughCountAggregation(plannerContext).rules()) // must run after PredicatePushDown and after TransformFilteringSemiJoinToInnerJoin
+                                .add(new RemoveFalseFiltersAfterDomainTranslator(plannerContext))
```

Comment:
RemoveTrivialFilters is not used here, but the new rule is. what's the selection criteria to when to use the new rule and when not to?

Reply:
Sorry, I've marked the PR as a draft because I was in the middle of the following process and since I wasn't done, this turned out to come with no explanation.

I tested the new rule by removing the "just for extra safety" `if`s from `PushPredicateIntoTableScan` and `RemoveFalseFiltersAfterDomainTranslator`. I noticed that `PushPredicateIntoTableScan` at `PlanOptimizers:600` may produce `None` predicates so I've added this new instance of `RemoveFalseFiltersAfterDomainTranslator`.

Then I got a cast exception :
```
Error:  io.trino.sql.planner.TestUnwrapCastInComparison.testNaN  Time elapsed: 0.088 s  <<< FAILURE!
io.trino.spi.TrinoException: Unable to cast double NaN to smallint
	at io.trino.type.DoubleOperators.saturatedFloorCastToLong(DoubleOperators.java:271)
	at io.trino.type.DoubleOperators.saturatedFloorCastToSmallint(DoubleOperators.java:249)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:757)
	at io.trino.sql.InterpretedFunctionInvoker.invoke(InterpretedFunctionInvoker.java:101)
	at io.trino.sql.InterpretedFunctionInvoker.invoke(InterpretedFunctionInvoker.java:53)
	at io.trino.sql.planner.DomainTranslator$Visitor.lambda$floorValue$7(DomainTranslator.java:915)
	at java.base/java.util.Optional.map(Optional.java:260)
	at io.trino.sql.planner.DomainTranslator$Visitor.floorValue(DomainTranslator.java:915)
	at io.trino.sql.planner.DomainTranslator$Visitor.coerceComparisonWithRounding(DomainTranslator.java:847)
	at io.trino.sql.planner.DomainTranslator$Visitor.visitComparisonExpression(DomainTranslator.java:559)
	at io.trino.sql.planner.DomainTranslator$Visitor.visitComparisonExpression(DomainTranslator.java:337)
...
	Suppressed: java.lang.Exception: Query: SELECT * FROM (VALUES CAST(NULL AS smallint)) t(a) WHERE a IS DISTINCT FROM nan() AND rand() = 42
...
```

So I've added a check of  
`typeCoercion.canCoerce(fromType, toType)` to `DomainTranslator$Visitor.floorValue` but then tests like
`TestDomainTranslator.testNumericTypeTranslation` failed because the check returned `false`.

I saw that sometimes we execute the coercion under a try-catch (for example at `UnwrapCastInComparison.java:312`) so I did the same (see commit 46e260df5324d8b8bb8a4419f0a44bff6196121b) and now the tests pass.

@findepi please let me know what you think about this approach.